### PR TITLE
[Fix] Rendertexture and shadowmap platform adaptation

### PR DESCRIPTION
**Pull request by GengineJS**

### CODE REVIEW

**Comment by GengineJS** [comment]
https://github.com/cocos-creator/engine/pull/8346

File: cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp [hunk 1]
```diff
@@ -458,11 +458,7 @@ void RenderAdditiveLightQueue::updateCameraUBO(const Camera *camera, gfx::Comman
     memcpy(uboCameraView.data() + UBOCamera::MAT_VIEW_PROJ_INV_OFFSET, camera->matViewProjInv.m, sizeof(cc::Mat4));
     TO_VEC3(uboCameraView, camera->position, UBOCamera::CAMERA_POS_OFFSET);
 
-    auto projectionSignY = device->getScreenSpaceSignY();
-    if (camera->getWindow()->hasOffScreenAttachments) {
-        projectionSignY *= device->getUVSpaceSignY(); // need flipping if drawing on render targets
-    }
-    uboCameraView[UBOCamera::CAMERA_POS_OFFSET + 3] = projectionSignY;
+    uboCameraView[UBOCamera::CAMERA_POS_OFFSET + 3] = (uint)(device->getScreenSpaceSignY() * 0.5 + 0.5) << 1 | (uint)(device->getClipSpaceSignY() * 0.5 + 0.5);
```
Comment (minggo) [inline cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: Can add a function to do `device->getScreenSpaceSignY() * 0.5 + 0.5) << 1 | (uint)(device->getClipSpaceSignY() * 0.5 + 0.5)`?

Comment (minggo) [reply cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: And please do not use c style cast.

Comment (GengineJS) [reply cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: > Can add a function to do `device->getScreenSpaceSignY() * 0.5 + 0.5) << 1 | (uint)(device->getClipSpaceSignY() * 0.5 + 0.5)`?

Yes, add it to the device?

Comment (GengineJS) [reply cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: > And please do not use c style cast.

Roger that

Comment (YunHsiao) [reply cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: Declare one inside the pipeline will do. Never attach irrelevant custom logic to the lower layer.

File: cocos/renderer/gfx-metal/MTLDevice.mm [hunk 2]
```diff
@@ -49,8 +49,8 @@ of this software and associated documentation files (the "Software"), to deal
 
 CCMTLDevice::CCMTLDevice() {
     _clipSpaceMinZ = 0.0f;
-    _screenSpaceSignY = 1.0f;
-    _UVSpaceSignY = 1.0f;
+    _screenSpaceSignY = -1.0f;
+    _clipSpaceSignY = 1.0f;
```
Comment (YunHsiao) [inline cocos/renderer/gfx-metal/MTLDevice.mm:53]: need to update this for vk device too

Comment (GengineJS) [reply cocos/renderer/gfx-metal/MTLDevice.mm:53]: > need to update this for vk device too

Yes, but I didn’t bring a windows computer today, so I’m not sure about the changes.

Comment (YunHsiao) [reply cocos/renderer/gfx-metal/MTLDevice.mm:53]: for vulkan, both clip and screen space sign Y should be -1

File: cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp (hunk 1 above)
Comment (GengineJS) [reply cocos/renderer/pipeline/RenderAdditiveLightQueue.cpp:461]: > Declare one inside the pipeline will do. Never attach irrelevant custom logic to the lower layer.

got it

File: cocos/renderer/gfx-metal/MTLDevice.mm (hunk 2 above)
Comment (GengineJS) [reply cocos/renderer/gfx-metal/MTLDevice.mm:53]: > for vulkan, both clip and screen space sign Y should be -1

I remember that I also changed other places, so the vulkan code has not been submitted yet

Comment (minggo) [reply cocos/renderer/gfx-metal/MTLDevice.mm:53]: > Yes, but I didn’t bring a windows computer today, so I’m not sure about the changes.

You can use Android device to test.

File: cocos/renderer/pipeline/forward/ForwardPipeline.h [hunk 3]
```diff
@@ -62,6 +62,7 @@ class CC_DLL ForwardPipeline : public RenderPipeline {
     void setSkybox(uint);
     void setShadows(uint);
     void destroyShadowFrameBuffers();
+    uint8_t getCombinationSignY();
```
Comment (minggo) [inline cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: getCombinationSignY -> getCombinationSignY const.
And `combinatonSignY` seems confusion.

Comment (GengineJS) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: confusion? Do you have a better suggestion? I will add an explanation to this function

Comment (minggo) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: It is not just a sign to indicate direction. It is more about to convert value from one region to another region. But i do not have better suggestion. May be can ask @YunHsiao or @ Jacksparrow2525 's help.

Comment (minggo) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: It is not just a sign to indicate direction. It is more about to convert value from one region to another region. But i do not have better suggestion. May be can ask @YunHsiao or @ Jacksparrow2525 's help.

Comment (GengineJS) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: I think combinatonSignY is okay, because it will be parsed on the shader side

Comment (minggo) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: k, please add description for it. And add const.

Comment (minggo) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: And i think it only needs to be computed once, not every frame.

Comment (GengineJS) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: Yes, I agree, so I will add a private variable for assignment

**Review by minggo — APPROVED** [review]

File: cocos/renderer/pipeline/forward/ForwardPipeline.h [hunk 4]
```diff
@@ -62,6 +62,7 @@ class CC_DLL ForwardPipeline : public RenderPipeline {
     void setSkybox(uint);
     void setShadows(uint);
     void destroyShadowFrameBuffers();
+    uint8_t getCombineSignY() const;
```
Comment (minggo) [inline cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: please add some comments for combineSignY

Comment (GengineJS) [reply cocos/renderer/pipeline/forward/ForwardPipeline.h:65]: combineSignY is commented on the js side

**Pull request merged**